### PR TITLE
Fix LLVM assertion errors for vector mode

### PR DESCRIPTION
**Pull request by vaithak**
took me a reasonable amount of time to figure out that vector mode also requires creating an overloaded function similar to gradient computation.

### CODE REVIEW

**Comment by vaithak** [comment]
clang-tidy is unable to post comments right now on any PR because of failing docker build, I have raised a PR to fix this in the original repo for the github action: https://github.com/ZedThree/clang-tidy-review/pull/90

File: test/ForwardMode/VectorMode.C [hunk 1]
```diff
@@ -12,8 +12,8 @@ double f1(double x, double y) {
 void f1_dvec(double x, double y, double *_d_x, double *_d_y);
 
 // CHECK: void f1_dvec(double x, double y, double *_d_x, double *_d_y) {
-// CHECK-NEXT:   clad::array<double> _d_vector_x = {1., 0.};
-// CHECK-NEXT:   clad::array<double> _d_vector_y = {0., 1.};
+// CHECK-NEXT:   clad::array<double> _d_vector_x = clad::one_hot_vector(2UL, 0UL);
```
Comment (parth-07) [inline test/ForwardMode/VectorMode.C:15]: `clad::one_hot_vector` and `clad::zero_vector` is a great addition. They have made the code
much more readable.

File: include/clad/Differentiator/Array.h [hunk 2]
```diff
@@ -246,6 +246,19 @@ template <typename T> class array {
   CUDA_HOST_DEVICE operator T*() const { return m_arr; }
 }; // class array
 
+// Function to instantiate a one_hot array of size n with 1 at index i
```
Comment (vgvassilev) [inline include/clad/Differentiator/Array.h:249]: What does `one_hot` mean?

Comment (vaithak) [reply include/clad/Differentiator/Array.h]: added a comment for this 👍🏼

Comment (vgvassilev) [reply include/clad/Differentiator/Array.h]: Why do we name it as a "hot" array?

File: include/clad/Differentiator/FunctionTraits.h [hunk 3]
```diff
@@ -37,130 +37,128 @@ namespace clad {
...
-  template <class ReturnType, class C, class... Args> 
-  struct return_type<ReturnType (C::*)(Args..., ...)> { 
-    using type = ReturnType; 
+  template <class ReturnType, class C, class... Args>
+  struct return_type<ReturnType (C::*)(Args..., ...)> {
+    using type = ReturnType;
   };
 
   // specializations for member functions pointer type with only cv-qualifiers
-  template <class ReturnType, class C, class... Args>
```
Comment (vgvassilev) [inline include/clad/Differentiator/FunctionTraits.h:74]: Can you configure your editor to not trim trailing whitespace and revert these changes?

File: include/clad/Differentiator/Array.h (hunk 2 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h]: This is a standard term: https://en.wikipedia.org/wiki/One-hot

File: include/clad/Differentiator/FunctionTraits.h (hunk 3 above)
Comment (vaithak) [reply include/clad/Differentiator/FunctionTraits.h]: reverted space changes from this file, a couple of minor changes remain in other files but I guess that shouldn't be an issue.

File: include/clad/Differentiator/Array.h (hunk 2 above)
Comment (vgvassilev) [reply include/clad/Differentiator/Array.h]: Ha, ok. There is always something new that I learn every day.

File: lib/Differentiator/VectorForwardModeVisitor.cpp [hunk 4]
```diff
@@ -138,28 +141,166 @@ VectorForwardModeVisitor::DeriveVectorMode(const FunctionDecl* FD,
...
+  DeclWithContext result =
+      m_Builder.cloneFunction(m_Function, *this, DC, m_Sema, m_Context, noLoc,
+                              vectorModeNameInfo, vectorModeFuncOverloadType);
+  FunctionDecl* vectorModeOverloadFD = result.first;
+
+  // Function declaration scope
+  beginScope(Scope::FunctionPrototypeScope | Scope::FunctionDeclarationScope |
+             Scope::DeclScope);
+  m_Sema.PushFunctionScope();
+  m_Sema.PushDeclContext(getCurrentScope(), vectorModeOverloadFD);
```
Comment (vgvassilev) [inline lib/Differentiator/VectorForwardModeVisitor.cpp:192]: We should probably extract these into some sort of a RAII object as it seems to be a common pattern throughout the code base. Perhaps we could do it in a separate PR.

File: include/clad/Differentiator/VectorForwardModeVisitor.h [hunk 5]
```diff
@@ -32,6 +32,11 @@ class VectorForwardModeVisitor : public BaseForwardModeVisitor {
   DerivativeAndOverload DeriveVectorMode(const clang::FunctionDecl* FD,
                                          const DiffRequest& request);
 
+  /// Builds an overload for the gradient function that has derived params for
+  /// all the arguments of the requested function and it calls the original
+  /// gradient function internally
```
Comment (vgvassilev) [inline include/clad/Differentiator/VectorForwardModeVisitor.h:37]: We should probably add an example to clarify what we mean here.

File: lib/Differentiator/VisitorBase.cpp [hunk 6]
```diff
@@ -179,11 +179,12 @@ namespace clad {
     return new (m_Context) DeclStmt(DGR, noLoc, noLoc);
   }
 
-  DeclRefExpr* VisitorBase::BuildDeclRef(DeclaratorDecl* D) {
+  DeclRefExpr* VisitorBase::BuildDeclRef(DeclaratorDecl* D,
+                                         const CXXScopeSpec* SS) {
```
Comment (vgvassilev) [inline lib/Differentiator/VisitorBase.cpp:183]:
```suggestion
                                         const CXXScopeSpec* SS/*=nullptr*/) {
```

That's something useful we have been doing in the codebase. I thought clang-tidy would suggest it..

File: lib/Differentiator/VisitorBase.cpp [hunk 7]
```diff
@@ -564,6 +564,30 @@ namespace clad {
...
+    clang::LookupResult R(m_Sema, declName, noLoc, Sema::LookupOrdinaryName);
+
+    // Find function declaration
+    NamespaceDecl* CladNS = GetCladNamespace();
+    CXXScopeSpec CSS;
+    CSS.Extend(m_Context, CladNS, loc, loc);
+    m_Sema.LookupQualifiedName(R, CladNS, CSS);
+
+    // Build the template specialization expression.
+    // TODO: currently this doesn't print func<templates>(args...) while dumping
```
Comment (vgvassilev) [inline lib/Differentiator/VisitorBase.cpp:581]:
```suggestion
    // FIXME: currently this doesn't print func<templates>(args...) while dumping
```

File: include/clad/Differentiator/VectorForwardModeVisitor.h (hunk 5 above)
Comment (vaithak) [reply include/clad/Differentiator/VectorForwardModeVisitor.h]: added example 👍🏼

**Review by vgvassilev — APPROVED** [review]
LGTM!
